### ISSUE TITLE
It shows black screen when loading many photos.

### ISSUE THREAD

**Issue by chongjae** [darickc/MMM-BackgroundSlideshow#118]
I loaded mounted folder from NAS.
It contains so many photos.

> magicmirror@2.18.0 start
> DISPLAY="${DISPLAY:=:0}" ./node_modules/.bin/electron js/electron.js

[20.04.2022 10:39.12.809] [LOG]   Starting MagicMirror: v2.18.0
[20.04.2022 10:39.12.821] [LOG]   Loading config ...
[20.04.2022 10:39.12.827] [LOG]   Loading module helpers ...
[20.04.2022 10:39.12.830] [LOG]   No helper found for module: alert.
[20.04.2022 10:39.12.831] [LOG]   No helper found for module: clock.
[20.04.2022 10:39.12.991] [LOG]   Initializing new module helper ...
[20.04.2022 10:39.12.992] [LOG]   Module helper loaded: calendar
[20.04.2022 10:39.12.994] [LOG]   No helper found for module: calendar_monthly.
[20.04.2022 10:39.13.423] [LOG]   Initializing new module helper ...
[20.04.2022 10:39.13.425] [LOG]   Module helper loaded: MMM-BackgroundSlideshow
[20.04.2022 10:39.13.426] [LOG]   All module helpers loaded.
[20.04.2022 10:39.13.516] [LOG]   Starting server on port 8080 ...
[20.04.2022 10:39.13.532] [LOG]   Server started ...
[20.04.2022 10:39.13.534] [LOG]   Connecting socket for: calendar
[20.04.2022 10:39.13.535] [LOG]   Starting node helper for: calendar
[20.04.2022 10:39.13.535] [LOG]   Connecting socket for: MMM-BackgroundSlideshow
[20.04.2022 10:39.13.537] [LOG]   Sockets connected & modules started ...
[20.04.2022 10:39.13.800] [LOG]   Launching application.
[20.04.2022 10:39.16.743] [INFO]  BACKGROUNDSLIDESHOW: Reading directory "modules/media/DCIM/Camera" for images.
[20.04.2022 10:42.38.969] [INFO]  BACKGROUNDSLIDESHOW: Reading directory "modules/media/Kidsnote" for images.
[20.04.2022 10:47.49.870] [INFO]  BACKGROUNDSLIDESHOW: **91573** files found

When load a little photos(about 1000~2000), It shows photos well.

But, when load so many photos, It just shows black screen.
How can I load it?

**Comment by MarianoMatute** [comment]
It is happening to me too. Loading 65K fotos never work the first attempt, restarting  for second time usally works.

**Comment by M4Rv** [comment]
It appears I have a similar problem, but I was only loading around 750 photos. At a random time after launching MM, usually around 2-3min. after launching, I get a black screen with only the mouse displayed. I think I narrowed to cause to be this module. The problem does not seem to occur when the module is disabled, or when the photo dir only contains one folder with only a couple of photos.

Some info about the system: I freshly installed the magicmirroros (docker) version yesterday.

magicmirror@2.29.0

### SYSTEM:   manufacturer: ; model: Docker Container; virtual: false
### OS:       platform: linux; distro: Debian GNU/Linux; release: 12; arch: arm64; kernel: 6.6.51+rpt-rpi-v8
### VERSIONS: electron: 31.6.0; used node: 20.17.0; installed node: 22.12.0; npm: 10.9.0; pm2:

**Comment by klalle** [comment]
@M4Rv I have 5k files, on a rpi 4, 2gb, and it never hangs since I switched over to firefox instead of electron. 
I know you run in docker... [looks](https://gitlab1s.com/khassel/magicmirror/-/blob/HEAD/build/entrypoint.sh#L177-L182) as if you pass the env arg `MM_SCENARIO=server` might start the docker in headless mode (without electron), and then you can start firefox from the rpi and see if that works? 
have a look at my [setup](https://github.com/darickc/MMM-BackgroundSlideshow/issues/148#issuecomment-2531583346) (not using docker)

**Comment by M4Rv** [comment]
I managed to fix my crashes by sanitizing my folders and files! This python script will duplicate my photos directory, clean out any special characters in folder and photo names, renames all photos to the name-of-the-parent-folder_number.jpg, and it will rescale large photos to fullhd resolution.
It requires pillow and pyheif. Feel free to try it out:

`import os
import shutil
from PIL import Image
import pyheif
import re

def sanitize_folder_name(name):
    translation_table = str.maketrans({"&": "und", "/": "", "\\": "", ":": "", "*": "", "?": "", "\"": "", "<": "", ">": "", "|": "", " ": "_"})
    return name.translate(translation_table)

def convert_heic_to_jpg(input_path, output_path):
    try:
        heif_file = pyheif.read(input_path)
        image = Image.frombytes(
            heif_file.mode, heif_file.size, heif_file.data, "raw", heif_file.stride
        )
        image.save(output_path, "JPEG")
        return True
    except Exception as e:
        return False

def resize_image(image_path, max_width, max_height):
    try:
        with Image.open(image_path) as img:
            img.thumbnail((max_width, max_height))
            img.save(image_path, "JPEG")
            return True
    except Exception as e:
        return False

def duplicate_and_process_folder(folder_path):
    base_folder = os.path.basename(folder_path)
    duplicate_folder = f"{base_folder}_copy"
    shutil.copytree(folder_path, duplicate_folder)

    errors = []

    for root, dirs, files in os.walk(duplicate_folder, topdown=False):
        for dir_name in dirs:
            sanitized_name = sanitize_folder_name(dir_name)
            if sanitized_name != dir_name:
                try:
                    os.rename(
                        os.path.join(root, dir_name), os.path.join(root, sanitized_name)
                    )
                except Exception as e:
                    errors.append(os.path.join(root, dir_name))

        parent_folder_name = sanitize_folder_name(os.path.basename(root))
        counter = 1

        for file_name in files:
            file_path = os.path.join(root, file_name)
            file_ext = os.path.splitext(file_name)[1].lower()
            new_file_name = f"{parent_folder_name}_{counter:04d}.jpg"
            new_file_path = os.path.join(root, new_file_name)

            if file_ext == ".heic":
                success = convert_heic_to_jpg(file_path, new_file_path)
                if success:
                    os.remove(file_path)
                else:
                    errors.append(file_path)

            elif file_ext == ".jpeg":
                try:
                    os.rename(file_path, new_file_path)
                except Exception as e:
                    errors.append(file_path)

            elif file_ext == ".jpg":
                try:
                    os.rename(file_path, new_file_path)
                except Exception as e:
                    errors.append(file_path)

            else:
                try:
                    with Image.open(file_path) as img:
                        img.convert("RGB").save(new_file_path, "JPEG")
                    os.remove(file_path)
                except Exception as e:
                    errors.append(file_path)

            if os.path.exists(new_file_path):
                resized = resize_image(new_file_path, 1920, 1080)
                if not resized:
                    errors.append(new_file_path)

            counter += 1

    return errors

if __name__ == "__main__":
    import sys
    if len(sys.argv) < 2:
        print("Usage: python script.py <folder_path>")
    else:
        folder_path = sys.argv[1]
        if os.path.isdir(folder_path):
            errors = duplicate_and_process_folder(folder_path)
            if errors:
                print("Errors encountered:")
                for error in errors:
                    print(error)
            else:
                print("Processing completed successfully.")
        else:
            print(f"The path '{folder_path}' is not a valid directory.")
`

**Comment by klalle** [comment]
@M4Rv so it wasn't the number of photos or even the resizing that caused the crash, but rather special characters in the names?

**Comment by klalle** [comment]
@M4Rv does the docker log show the output of MM? in that case you should get the error using `docker logs <nameofthedocker>` 

Took a look in the [entrypoint.sh](https://gitlab1s.com/khassel/magicmirror/-/blob/HEAD/build/entrypoint.sh#L178) that seems to be where MM is started from, and it doesn't look like the output are thrown to dev null anyway....

**Comment by M4Rv** [comment]
@klalle yes, that is actually where i was looking. I see all the info, warn and debug log entries there, but there was no error logged on that crash scenario (at least when using MMM-BackgroundSlideShow - with the MMM-RandomPhoto I got some stacktrace in the log).

**Comment by klalle** [comment]
@M4Rv didn't you say earlier that the backend was working fine (you could start a browser and access the MM even though the electron app had stopped responding)? 
in that case, the error must be in the frontend output (f12, or maybe Ctrl + Shift + I for an electron app?)
